sys/kern/exec.c

```diff
@@ -323,6 +323,25 @@ static bool check_setid(vnode_t *vn, uid_t *uid, gid_t *gid) {
   return (*uid != (uid_t)-1) || (*gid != (gid_t)-1);
 }
 
+static char *prepare_pargs(exec_args_t *args) {
+  char *pargs = kmalloc(M_STR, MAX_PARGS_LEN, M_ZERO);
+  size_t it = 1, used = 0, max = MAX_PARGS_LEN;
+
+  while (used < max && it < args->argc) {
+    size_t len = strlcpy(pargs + used, args->argv[it], max - used);
+
+    /* calculate how much we have copied */
+    len = min(len, max - used);
+    used += len;
+
+    /* add space between args */
+    if (used < max)
+      pargs[used++] = ' ';
+    ++it;
+  }
```

Comment:
You need to be careful about `used` and `max`: you need to decide if `used` includes the terminating null byte or not.
Assuming it does not include it, the loop should look like this:
```suggestion
  while (used + 1 < max && it < args->argc) {
    size_t len = strlcpy(pargs + used, args->argv[it], max - used);

    /* calculate how much we have copied */
    len = min(len, max - used - 1);
    used += len;

    /* add space between args */
    if (used + 1 < max)
      pargs[used++] = ' ';
    ++it;
  }
```

Reply:
Thanks